Bug 1212950 - EAP 6 start operation causes agent to run out of memory du...

lzoubek opened the pull request:
...e to

storing console output in an unused buffer

Now process output is captured (if captured) up to 2MB size, once output
exceeds this limit, it is ignored - so we don't run out of memory (unless
agent does not start plenty of verbose processes). Default limit can be
changed via rhq.process-execution.captured-output.limit system property.

This commit also gives more power to plugin writers about capturing process
outputs. ProcessExecution#setCaptureOutput is now deprecated in favor of new
CaptureMode setting. CaptureMode can capture to memory and/or forward to
agent.log as well as setting captured limit.

tsegismont commented:
@lzoubek thanks for the PR. I'll try to review it today.

tsegismont commented on modules/core/native-system/src/main/java/org/rhq/core/system/ProcessExecutionOutputStream.java:
```diff
@@ -0,0 +1,103 @@
+        }
+        DEFAULT_LIMIT = limit;
+    }
+
```
I think a call to `java.lang.Integer#getInteger(java.lang.String, int)` would be easier to read (and does the same thing).
```java
private static final int DEFAULT_LIMIT = Integer.getInteger("rhq.process-execution.captured-output.limit", 2 * 1024 * 1024);  // default 2MB
```

tsegismont commented on modules/core/native-system/src/main/java/org/rhq/core/system/ProcessExecutionOutputStream.java:
```diff
@@ -0,0 +1,103 @@
+     * sets new buffer limit
+     * @param limit in Bytes
+     */
+    public synchronized void setLimit(int limit) {
```
In my opinion, `writeToLog` and `limit` should be set when creating the instance and not be modifiable. It does not really make sense to change them once the instance is created.

tsegismont commented:
Other than the two comments inline, the change looks good to me. Could you replace calls to the deprecated method in the codebase? Or at least in the EAP 6 plugin ?

lzoubek commented:
@tsegismont thank you for reviewing, I've fixed and rebased my original commit. I'll replace deprecated calls in EAP6 plugin within a separate commit, I need to make sure it does not get to 3.3.x branch.

tsegismont commented:
OK. I understand the concern, but please make sure to mention that in the master commit otherwise if we get into any trouble then we'll spend some time before figuring out that part of the code is not in the 3.3.x branch.
